FBX Importer - Fixes for 3.2 [WIP]

RevoluPowered opened the pull request:
Skinning support added

Various fixes
- Fixes bind mount id and mesh index
- Fixed duplicate nodes being created
- Prevented leak when instances being freed during re-import.
- Improved camera and light transform import
- skeleton handling and technical debt removal
- ASSIMP: bone nodes were unlinked from bones by this code
- bone_add working can distinguish between armatutes
- Updated transform to be the correct offset
- Added safety for state.root node errors
- Fixed memory leak with leaf bones
- Implemented children reparenting for mesh template
- Fixed godot import segfault
- Fix build failing on mono
- Clear resources we use which are no longer required after import
- Fixed bone duplication issue
- Working skeleton_bone_map which can lookup armatures properly now.
- **No more guessing armatures** (enhancement: assimp provides them to us - much less code)
- Animation names are filtered for UTF8 safety due to import issue with UTF8 chars.
- No longer need armature stack removed it
- Fixed stack being used up when mesh swapped & Fixed bone ID

**Notes:**
We use a mesh template which is a fake node to instance the initial
mesh nodes. This is to ensure the entire tree can be built.

We replace mesh node templates with the real mesh after the
skeleton is available, since this makes it ensure that the fully
built skeleton exists with all bones, all nodes, etc.

- Improved animation skeleton lookup

This uses bone stack to find the relevant bones.

The bone stack is a stack which pops when it finds bones,
this overcomes duplicate bones with the same names.

FBX has lots of these because animation armature has bone names like bone001
and another armature will also have bone001

**Assimp Pull Requests - Dependencies**
- clang formatting in assimp https://github.com/assimp/assimp/pull/2728
- easy armatures, multi root from assimp https://github.com/assimp/assimp/pull/2731
- _put animation truncation pr here_
- _put pivot handling pr here_
- _put single pr for FBX keyframe dictionary lookup here_ https://github.com/assimp/assimp/pull/2729/commits/4910587585896a057b1cb3ced8dfa4583150871b

I am still separating my PRs up as there is a lot of changes which are being merged soon.

RevoluPowered committed:
FBX Importer Generation 3

Basic skin support

Various fixes

- Fixes bind mount id and mesh index
- Fixed duplicate nodes being created
- Prevented leak when instances being freed during re-import.
- Improved camera and light transform import
- skeleton handling and technical debt removal
- ASSIMP: bone nodes were unlinked from bones by this code
- bone_add working can distinguish between armatutes
- Updated transform to be the correct offset
- Added safety for state.root node errors
- Fixed memory leak with leaf bones
- Implemented children reparenting for mesh template

We use a mesh template which is a fake node to instance the initial
mesh nodes . This is to ensure the entire tree can be built.

We replace mesh node templates with the real mesh after the
skeleton is available, since this makes it ensure that the fully
built skeleton exists with all bones, all nodes, etc.

- Improved animation skeleton lookup

This uses bone stack to find the relevant bones.

The bone stack is a stack which pops when it finds bones,
this overcomes duplicate bones with the same names.

FBX has lots of these because animation armature has bone names like bone001
and another armature will also have bone001

RevoluPowered committed:
Fixed errors in node path assignment

This means that the node path and node list are updated.

Before the animation handler finds them this means that the node path
is correct.

Removed node path not required.

akien-mga commented:
You're re-adding commits which are already in the `master` branch, a rebase is likely needed.

RevoluPowered commented:
@akien-mga going to rebase on master just after I am finished my crunch today.

RevoluPowered committed:
import_animation fixes to basic skeleton data

Adds some more debug messages
Fixed godot import segfault
Fix build failing on mono
Clear resources we use which are no longer required after import
Fixed bone duplication issue

Working skeleton_bone_map which can lookup armatures properly now.

WIP fixes for multi armature support


Fixes in this version:
- Multiple armature high level algorithm
- No more guessing armatures (enhancement: assimp provides them to us - much less code)
- Animation names are filtered for UTF8 safety due to import issue with UTF8 chars.
- No longer need armature stack removed it

Fixed stack being used up when mesh swapped & Fixed bone ID

Simple explanation:
- Every mesh uses a node from the stack
- Node stack was empties before completed
- Every time node not found, stack must be rebuilt to maintain correct armature order :)

Additional fixes:
- Fixes destructor in assimp
- Implements aiNode* mArmature in bone data
- Implements aiNode* mParent in bone data
- Fixes parent ID on bones.

RevoluPowered committed:
assimp update to dev version

RevoluPowered committed:
Implemented skeleton assignment in generate_mesh_indicies

This is the only place we can safely do a lookup for the skeleton for the mesh.h
I used a pointer reference so we can pass this back out, since the skeleton assignment happens inside the function.

Interactive rebase bug - broken some things - cross comparing

Merge remote-tracking branch 'refs/remotes/origin/feature/generation3importer'

Conflicts:
	drivers/gles2/shaders/scene.glsl
	drivers/gles3/shaders/scene.glsl
	editor/import/editor_scene_importer_gltf.cpp
	editor/spatial_editor_gizmos.cpp
	modules/assimp/editor_scene_importer_assimp.cpp
	scene/3d/skeleton.cpp
	scene/3d/skeleton.h
	scene/resources/skin.cpp
	scene/resources/skin.h
	thirdparty/assimp/code/FBX/FBXConverter.cpp

RevoluPowered committed:
Fixed set_bind_bone call missing skin was not refering to the correct bones

note: this shows skin reuse is not working
yet but is a good step in the right direction

RevoluPowered committed:
Added mesh reparenting to the armature node this is a permanent feature and must be enforced

RevoluPowered commented:
Going to rebase in a way to just have 'added skinning support commit' and a single commit updating assimp.

RevoluPowered committed:
Fixed import_animation spawning tracks per skin

This should only be per bone in the skeleton

RevoluPowered committed:
WIP

RevoluPowered committed:
WIP pivot code cleanup

RevoluPowered committed:
Added global transform to pivot calculation

RevoluPowered committed:
Removing old handler for node chain

This was needed for creating fake pivot nodes

RevoluPowered committed:
Goodbye node chain handler

RevoluPowered committed:
Node creation fix

RevoluPowered committed:
Removing inverse pivot data - redundant

Adding safety for blender files

RevoluPowered committed:
Added helper for assimp vector

RevoluPowered committed:
Removing dead code for anims and identity pivot points

RevoluPowered committed:
wip pivot plumbing in animations - no crashes

RevoluPowered committed:
Added Geometric transform 'support'

However we need the geometiric transforms to be ignored on children meaning we need to inverse every geometric child on each child :)

Later on we can add this to the nodes

RevoluPowered committed:
disabling prints

RevoluPowered committed:
Added debug message

RevoluPowered committed:
Adding pivot status

RevoluPowered committed:
Refactoring and disabling pivot animation filtering in 2840
